src/hotspot/share/opto/memnode.cpp

```diff
@@ -3519,17 +3519,130 @@ Node *StoreNode::Ideal_masked_input(PhaseGVN *phase, uint mask) {
 // (StoreB ... (RShiftI _ (LShiftI _ valIn conIL ) conIR) )
 // If (conIL == conIR && conIR <= num_bits)  this simplifies to
 // (StoreB ... (valIn) )
-Node *StoreNode::Ideal_sign_extended_input(PhaseGVN *phase, int num_bits) {
-  Node *val = in(MemNode::ValueIn);
-  if( val->Opcode() == Op_RShiftI ) {
-    const TypeInt *t = phase->type( val->in(2) )->isa_int();
-    if( t && t->is_con() && (t->get_con() <= num_bits) ) {
-      Node *shl = val->in(1);
-      if( shl->Opcode() == Op_LShiftI ) {
-        const TypeInt *t2 = phase->type( shl->in(2) )->isa_int();
-        if( t2 && t2->is_con() && (t2->get_con() == t->get_con()) ) {
-          set_req_X(MemNode::ValueIn, shl->in(1), phase);
-          return this;
+// If (conIL > conIR) we are inventing 0 lower bits, and throwing
+// away upper bits, but we are not introducing garbage bits from the upper bits
+// (the right in the notation below).
+// We can simplify into
+// (StoreB ... (LShiftI _ valIn (conIL - conIR)) )
+// This case happens when the right-hand side of the store was itself a left shift, that gets merged
+// into the inner left shift of the sign-extension. For instance, if we have
+// array_of_shorts[0] = (short)(X << 2)
+// We get a structure such as:
+// (StoreB ... (RShiftI _ (LShiftI _ (LShiftI _ X 2) 16) 16))
+// that is simplified into
+// (StoreB ... (RShiftI _ (LShiftI _ X 18) 16)).
+// It is thus useful to handle the case where conIL > conIR.
+//
+// Let's assume we have the following 32 bits integer that we want to stuff in 8 bits char:
+// +------------------------+---------+
+// |        v[8..31]        | v[0..7] |
+// +------------------------+---------+
+//  31                     8 7        0
+// v[0..7] is meaningful, but v[8..31] is not. In this case, num_rejected_bits == 24.
```

Comment:
Would this example not be nice with the original case above?
```
// Check for useless sign-extension before a partial-word store
// (StoreB ... (RShiftI _ (LShiftI _ valIn conIL ) conIR) )
// If (conIL == conIR && conIR <= num_bits)  this simplifies to
// (StoreB ... (valIn) )
```
Because it seems you are assuming here that `conIL == conIR`, right? And then below you ask what if they are not equal.

Reply:
It could also be nice to introduce the `num_rejected_bits` somewhere.

Reply:
I don't understand what you mean. I'm not speaking about `conIR` and `conIL` here. I'm just saying we have a value v that we want to store in a 8-bit long storage (a byte). We can see v as made of its 8 lower bits (that we want to actually store) and the 24 upper bits (that we don't care about). I'm just introducing notations, but I haven't done any operation yet. I tried to rephrase around there.